File: A0/pass.f90
program ping
    implicit none
    include 'mpif.h'

    integer :: numtasks, rank, dest, source,count, tag, ierr
    integer :: stat(MPI_STATUS_SIZE)   !required variable for receive routines
    integer :: inmsg,outmsg
    outmsg = 0
    tag = 1
    call MPI_INIT(ierr)
    call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)

    ! task 0 sends to task 1 and waits to receive a return message
    if (rank==0) then
    write(*,*)
    Write(*,'(A)') 'Testing MPI Message Passing...'
        dest = 1
        source = 1
    call MPI_SEND(outmsg, 1, MPI_INTEGER, dest, tag, MPI_COMM_WORLD, ierr)
    call MPI_RECV(inmsg, 1, MPI_INTEGER, source, tag, MPI_COMM_WORLD, stat,ierr)

    ! task 1 waits for task 0 message then returns a message
    else if (rank==1) then
        dest = 0
        source= 0 
        call MPI_RECV(inmsg, 1, MPI_INTEGER, source, tag, MPI_COMM_WORLD, stat,ierr)
        outmsg=inmsg+1
        call MPI_SEND(outmsg, 1, MPI_INTEGER, dest, tag, MPI_COMM_WORLD, ierr)
    endif

    ! query recieve Stat variable and print message details
    call MPI_GET_COUNT(stat, MPI_CHARACTER, count, ierr)
    if(rank==0) then
        if(inmsg==1) then
            WRITE(*,'(A)') 'Message Pass Succeeded'
        else
            WRITE(*,'(A)') 'Message Pass Failed'
        endif
        write(*,*)
    endif
    !WRITE(*,*)  'Task ',rank,': Received', count, 'char(s) from task', stat(MPI_SOURCE),'with tag',stat(MPI_TAG)

    call MPI_FINALIZE(ierr)

end program ping

File: A0/pass.c
#include "mpi.h"
#include <stdio.h>

int main(int argc, char *argv[])  {
    int numtasks, rank, dest, source, rc, count, tag=1; 
    int outmsg,inmsg;
    MPI_Status Stat;   // required variable for receive routines

    outmsg=0;

    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD, &numtasks);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    // task 0 sends to task 1 and waits to receive a return message
    if (rank == 0) {
        printf("\nTesting MPI Message Passing...\n");
        dest = 1;
        source = 1;
        MPI_Send(&outmsg, 1, MPI_INT, dest, tag, MPI_COMM_WORLD);
        MPI_Recv(&inmsg, 1, MPI_INT, source, tag, MPI_COMM_WORLD, &Stat);
    } 
                            
    // task 1 waits for task 0 message then returns a message
    else if (rank == 1) {
        dest = 0;
        source = 0;
        MPI_Recv(&inmsg, 1, MPI_INT, source, tag, MPI_COMM_WORLD, &Stat);
        outmsg=inmsg+1;
        MPI_Send(&outmsg, 1, MPI_INT, dest, tag, MPI_COMM_WORLD);
    }
                            
    // query recieve Stat variable and print message details
    MPI_Get_count(&Stat, MPI_CHAR, &count);
    if(rank==0){
        if(inmsg==1){
            printf("Message Pass Succeeded\n\n");
        }else{
            printf("Message Pass Failed\n\n");
        }
    }
    //printf("Task %d: Received %d char(s) from task %d with tag %d \n",rank, count, Stat.MPI_SOURCE, Stat.MPI_TAG);
                            
    MPI_Finalize();
}
                            
